src/core/EdFi.DataManagementService.Core/Backend/PostgresqlDocumentStoreRepository.cs

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System.Text.Json.Nodes;
+using Dapper;
+using EdFi.DataManagementService.Core.Model;
+using Microsoft.Extensions.Logging;
+using Npgsql;
+
+namespace EdFi.DataManagementService.Core.Backend;
+
+public class PostgresqlDocumentStoreRepository(
+    ILogger<PostgresqlDocumentStoreRepository> _logger,
+    string _connectionString
+) : IDocumentStoreRepository, IQueryHandler
+{
+    // Returns an integer in the range 0..15 from the last byte of a DocumentUuid
+    private static int PartitionKeyFor(DocumentUuid documentUuid)
+    {
+        Guid asGuid = Guid.Parse(documentUuid.Value);
+        byte lastByte = asGuid.ToByteArray()[^1];
+        return lastByte % 16;
+    }
+
+    public async Task<UpsertResult> UpsertDocument(UpsertRequest upsertRequest)
+    {
+        _logger.LogDebug("Entering UpsertDocument - {TraceId}", upsertRequest.TraceId);
+
+        await using (var conn = new NpgsqlConnection(_connectionString))
+        {
+            conn.Open();
+            var command =
+                $"INSERT INTO public.documents(document_partition_key, document_uuid, resource_name, edfi_doc) "
+                + $"VALUES ({PartitionKeyFor(upsertRequest.DocumentUuid)}, '{upsertRequest.DocumentUuid.Value}', '{upsertRequest.ResourceInfo.ResourceName.Value}', '{upsertRequest.EdfiDoc.ToJsonString()}');";
```

Comment:
Would it make sense to use Dapper anonymous parameters here, for safer SQL statements?

Reply:
Definitely. Addressed.